src/lib.rs

```diff
@@ -0,0 +1,281 @@
+// modified, or distributed except according to those terms.
+
+//! Create a [`Sink`] implementation from an initial value and a closure
+//! returning a [`Future`].
+//!
+//! This is very similar to how `futures::stream::unfold` creates a `Stream`
+//! implementation from a seed value and a future-returning closure.
+//!
+//! # Examples
+//!
+//! ```
+//! use async_std::io;
+//! use futures::prelude::*;
+//! use quicksink::Action;
+//!
+//! quicksink::make_sink(io::stdout(), |mut stdout, action| async move {
+//!     match action {
+//!         Action::Send(x) => stdout.write_all(x).await?,
+//!         Action::Flush => stdout.flush().await?,
+//!         Action::Close => stdout.close().await?
+//!     }
+//!     Ok::<_, io::Error>(stdout)
+//! });
+//! ```
+
+use futures_core::ready;
+use futures_sink::Sink;
+use std::{future::Future, pin::Pin, task::{Context, Poll}};
+
+/// Returns a `Sink` impl based on the initial value and the given closure.
+///
+/// The closure will be applied to the initial value and an [`Action`] that
+/// informs it about the action it should perform. The returned [`Future`]
+/// will resolve to another value and the process starts over using this
+/// output.
+pub fn make_sink<S, F, T, A, E>(init: S, f: F) -> SinkImpl<S, F, T, A, E>
+where
+    F: FnMut(S, Action<A>) -> T,
+    T: Future<Output = Result<S, E>>,
+{
+    SinkImpl {
+        lambda: f,
+        future: None,
+        param: Some(init),
+        state: State::Empty,
+        _mark: std::marker::PhantomData
+    }
+}
+
+/// The command given to the closure so that it can perform appropriate action.
+///
+/// Presumably the closure encapsulates a resource to perform I/O. The commands
+/// correspond to methods of the [`Sink`] trait and provide the closure with
+/// sufficient information to know what kind of action to perform with it.
+#[derive(Debug)]
+pub enum Action<A> {
+    /// Send the given value.
+    /// Corresponds to [`Sink::start_send`].
+    Send(A),
+    /// Flush the resource.
+    /// Corresponds to [`Sink::poll_flush`].
+    Flush,
+    /// Close the resource.
+    /// Corresponds to [`Sink::poll_close`].
+    Close
+}
+
+/// The various states the `Sink` may be in.
+#[derive(Debug, PartialEq, Eq)]
+enum State {
+    /// The `Sink` is idle.
+    Empty,
+    /// The `Sink` is sending a value.
+    Sending,
+    /// The `Sink` is flushing its resource.
+    Flushing,
+    /// The `Sink` is closing its resource.
+    Closing,
+    /// The `Sink` is closed (terminal state).
+    Closed
+}
+
+/// `SinkImpl` implements the `Sink` trait.
+#[derive(Debug)]
+pub struct SinkImpl<S, F, T, A, E> {
+    lambda: F,
+    future: Option<T>,
+    param: Option<S>,
+    state: State,
+    _mark: std::marker::PhantomData<(A, E)>
+}
+
+impl<S, F, T, A, E> SinkImpl<S, F, T, A, E> {
+    /// Project the closure from the pinned struct.
+    fn lambda(self: Pin<&mut Self>) -> &mut F {
+        unsafe {
```

Comment:
Couldn't you also do this with https://github.com/taiki-e/pin-project without `unsafe`?

Reply:
I went with https://github.com/taiki-e/pin-project-lite to minimise dependencies.